From d09cb3bb713433f74066d8dce7b38932c206c139 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marek=20Marczykowski-G=C3=B3recki?=
 <marmarek@invisiblethingslab.com>
Date: Thu, 5 Feb 2015 05:59:30 +0100
Subject: [PATCH 12/16] libxl: pass ipaddr to libxl toolstack
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
Organization: Invisible Things Lab
Cc: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>

Do not silently ignore its value. LibXL support only one address, so
refuse multiple IPs.

Signed-off-by: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>
---
 src/libxl/libxl_conf.c   | 11 +++++++++++
 src/libxl/libxl_domain.c | 12 ++++++++++++
 2 files changed, 23 insertions(+)

diff --git a/src/libxl/libxl_conf.c b/src/libxl/libxl_conf.c
index 6d3c443..f08b1b8 100644
--- a/src/libxl/libxl_conf.c
+++ b/src/libxl/libxl_conf.c
@@ -1087,6 +1087,11 @@ libxlMakeNic(virDomainDefPtr def,
         case VIR_DOMAIN_NET_TYPE_ETHERNET:
             if (VIR_STRDUP(x_nic->script, l_nic->script) < 0)
                 return -1;
+            if (l_nic->nips > 0) {
+                x_nic->ip = virSocketAddrFormat(&l_nic->ips[0]->address);
+                if (!x_nic->ip)
+                    return -1;
+            }
             break;
         case VIR_DOMAIN_NET_TYPE_NETWORK:
         {
@@ -1104,6 +1109,12 @@ libxlMakeNic(virDomainDefPtr def,
                 return -1;
             }
 
+            if (l_nic->nips > 0) {
+                x_nic->ip = virSocketAddrFormat(&l_nic->ips[0]->address);
+                if (!x_nic->ip)
+                    return -1;
+            }
+
             if ((brname = virNetworkGetBridgeName(network))) {
                 if (VIR_STRDUP(x_nic->bridge, brname) < 0)
                     fail = true;
diff --git a/src/libxl/libxl_domain.c b/src/libxl/libxl_domain.c
index 856cfb4..f0eaf6c 100644
--- a/src/libxl/libxl_domain.c
+++ b/src/libxl/libxl_domain.c
@@ -482,6 +482,18 @@ libxlDomainDeviceDefPostParse(virDomainDeviceDefPtr dev,
         STRNEQ(def->os.type, "hvm"))
         dev->data.chr->targetType = VIR_DOMAIN_CHR_CONSOLE_TARGET_TYPE_XEN;
 
+    if (dev->type == VIR_DOMAIN_DEVICE_NET &&
+            (dev->data.net->type == VIR_DOMAIN_NET_TYPE_BRIDGE ||
+             dev->data.net->type == VIR_DOMAIN_NET_TYPE_ETHERNET ||
+             dev->data.net->type == VIR_DOMAIN_NET_TYPE_NETWORK)) {
+        if (dev->data.net->nips > 1) {
+            virReportError(VIR_ERR_CONFIG_UNSUPPORTED,
+                    _("multiple IP addresses not supported on device type %s"),
+                    virDomainNetTypeToString(dev->data.net->type));
+            return -1;
+        }
+    }
+
     if (dev->type == VIR_DOMAIN_DEVICE_HOSTDEV ||
         (dev->type == VIR_DOMAIN_DEVICE_NET &&
          dev->data.net->type == VIR_DOMAIN_NET_TYPE_HOSTDEV)) {
-- 
1.8.3.1

